Xamarin.Forms.Exceptions/XamlParseException.cs

```diff
@@ -0,0 +1,123 @@
+		readonly string _message;
+		readonly string _undefinedMessage = Resources.UndefinedException;
+		static readonly string PositionMessage = Resources.Position;
+
+		string GetTypeString(bool isError) => isError ? Resources.Error : Resources.Warning;
+
+		public virtual string Prefix { get; }
+
+		public bool Error { get; private set; }
+
+		// http://blogs.msdn.com/b/msbuild/archive/2006/11/03/msbuild-visual-studio-aware-error-messages-and-message-formats.aspx
+		public virtual string GetMessage()
+		{
+			throw new NotImplementedException();
+		}
+
+		public IXmlLineInfo XmlInfo { get; }
+
+		public int Code { get; }
+
+		public string ErrorCode => $"{Prefix}{Code:0000}";
+
+		public virtual string ErrorSubcategory { get; }
+
+		public virtual string HelpKeyword { get; }
+
+#if NETSTANDARD2_0
+		protected XamlParseException(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context)
+			: base(info, context)
+		{
+			if (int.TryParse(info.GetString("errorCode"), out int code))
+				Code = code;
+			ErrorSubcategory = info.GetString("errorCode");
+			HelpKeyword = info.GetString("helpKeyword");
+		}
+
+		[System.Security.Permissions.SecurityPermission(System.Security.Permissions.SecurityAction.Demand, SerializationFormatter = true)]
+		public override void GetObjectData(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context)
+		{
+			base.GetObjectData(info, context);
+			info.AddValue("errorSubcategory", ErrorSubcategory);
+			info.AddValue("errorCode", ErrorCode);
+			info.AddValue("helpKeyword", HelpKeyword);
+		}
+#endif
+
+		public XamlParseException(int code, params string[] args)
+			: this(code, xmlInfo: null, args)
+		{
+		}
+
+		protected XamlParseException(int code, IXmlLineInfo xmlInfo, params string[] args)
+			: this(code, xmlInfo, innerException: null, args)
+		{
+		}
+
+		public XamlParseException(int code, IXmlLineInfo xmlInfo, Exception innerException, params string[] args)
+			: this(code, true, xmlInfo, innerException, args)
+		{
+		}
+
+		public XamlParseException(int code, bool error, IXmlLineInfo xmlInfo, Exception innerException, params string[] args)
+			: base(code.ToString("0000"), innerException)
+		{
+			Code = code;
+			XmlInfo = xmlInfo;
+			Error = error;
+			_message = Format(GetMessage(), args);
+		}
+
+		// http://blogs.msdn.com/b/msbuild/archive/2006/11/03/msbuild-visual-studio-aware-error-messages-and-message-formats.aspx
+		string Format(string unformattedMessage, params object[] args)
+		{
+			var message = string.IsNullOrEmpty(unformattedMessage)
+				? _undefinedMessage
+				: $"{GetTypeString(Error)} {Prefix}{Code:0000}: {SmartFormat(unformattedMessage, args)}";
+
+			return AddPosition(XmlInfo, message);
+		}
+
+		string SmartFormat(string formatString, params object[] args)
+		{
+			var matches = _curlyBracketRegex.Matches(formatString);
```

Comment:
what's the point of using a regex here ? I think we can rely on string format to do the argument validation